clippy_lints/src/map_nil_fn.rs

```diff
@@ -0,0 +1,187 @@
+use rustc::hir;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::codemap::Span;
+use utils::{in_macro, iter_input_pats, match_type, method_chain_args, snippet, span_lint_and_then};
+use utils::paths;
+
+#[derive(Clone)]
+pub struct Pass;
+
+/// **What it does:** Checks for usage of `Option.map(f)` where f is a nil
+/// function or closure
```

Comment:
Whoops, thanks for the explanation. I will tidy up the user facing portions to consistently use `unit`. I added the divergent support just for completeness and I couldn't think of any usecases apart from maybe `_.map(std::process::exit)`. I can remove it to simplify the code slightly